### PR TITLE
feat(domainstorytelling): add domain storytelling diagram

### LINKED ISSUE #7942: Diagram Proposal: Domain Storytelling
## Proposal

Mermaid.js should add **Domain Storytelling** as a new diagram type, because it fills a
gap in Mermaid's coverage of collaborative and Domain-Driven Design (DDD) modeling.

Domain Storytelling is a collaborative modeling technique created by Stefan Hofer and
Henning Schwentner (see [domainstorytelling.org](https://domainstorytelling.org)). Domain
experts tell short stories about how they work, and the team records them as a sequence of
sentences following a fixed grammar:

> **an actor** *does something* **with a work object** *and passes it to* **another actor**.

Each story is a numbered sequence of these sentences, so a reader can literally read the
diagram back as prose. It is a "picture the domain" technique that serves as a natural
on-ramp to bounded-context discovery, Event Storming, and Context Mapping.

Today Mermaid can express *structure* (class, ER, C4-style relationships) and *technical
flow* (sequence, flowchart), but it has no first-class way to capture **who does what with
which business object, in what order** at the language of the domain expert. Domain
Storytelling complements the existing diagrams rather than overlapping them: it is
actor/work-object oriented, ordered, and deliberately non-technical.

Bringing it into Mermaid's text-based DSL means domain stories become **diff-able,
reviewable, and version-controlled alongside the code they describe** — instead of living in
a separate proprietary modeling tool or a screenshot in a wiki.

## Use Cases

- **Requirements & discovery workshops** — capture a business process live with domain
  experts, in their own words, without dropping to a technical flowchart.
- **Domain-Driven Design** — use the actors/work objects/groups to surface **bounded
  contexts** and hand off into Context Mapping / Event Storming.
- **Onboarding & knowledge sharing** — a numbered, readable story explains a process to new
  team members in the domain's own vocabulary.
- **Documentation-as-code** — keep the domain story next to the service it models, in the
  same repo, reviewed in the same PR.
- **Living documentation** — because it is text, the story can be updated and reviewed as
  the process evolves.
- **Bridging business and engineering** — one artifact both a product owner and an engineer
  can read.

## Screenshots

Domain Storytelling diagrams render actors and work objects as labeled icon nodes connected
by **numbered activity arrows** that form a readable sentence sequence, with optional
**groups** and **annotations**.

a sample from the Domain Storytelling literature (ELK and Dagre):
Image
Image

Annotation Example:
Image

## Syntax

Every diagram starts with `domainstorytelling-beta`. The core sentence is:

```
A_Actor : 01 -- "activity" -> W_WorkObject
```

Identifiers carry a type prefix — `A_` actors, `W_` work objects, `G_` groups,
`S_` optional sentence ids — and sentences are numbered (`01`, `02`, …) to define the story
order.

**Basic story:**

```
domainstorytelling-beta

A_Customer : 01 -- "places" -> W_Order
A_Service  : 02 -- "processes" -> W_Order
```

**Declarations (labels, icons, group membership):**

```
domainstorytelling-beta

group G_FrontOffice "Front Office"

A_Customer "Customer" person in G_FrontOffice
A_Clerk    "Service Clerk" person in G_FrontOffice
W_Order    "Purchase Order" document

A_Customer : 01 -- "places" -> W_Order
A_Clerk    : 02 -- "reviews" -> W_Order
```

**Icons (optional).** Mermaid ships a built-in icon pack covering the base notation, so
these render with no setup and are referenced by bare name: `person`, `people`, `system`
for actors and `document`, `folder`, `call`, `email`, `conversation`, `info` for work
objects. Icons are optional per node — without a declaration, `person` is applied to actors
and `document` to work objects. Beyond the built-in set, any
[Iconify](https://iconify.design) icon can be used in `pack:icon` form (for example
`fa6-solid:user`), with the pack registered via `mermaid.registerIconPacks(...)`.

**Chained continuations (multiple work objects / actors, reverse links):**

```
domainstorytelling-beta

A_Sales : 01 -- "hands over" -> W_QualifiedLead -- "to" -> A_Consulting -- "and" -> A_Onboarding
A_ProductOwner : 02 -- "aligns on" -> W_Roadmap <- "provides feedback on" -- A_Customer
```

**Groups (bounded contexts), incl. nesting and block form:**

```
domainstorytelling-beta

group G_BackOffice "Back Office" {
  A_Clerk     : 01 -- "creates" -> W_Order
  A_Warehouse : 02 -- "packs" -> W_Package
}
group G_Logistics "Logistics"
A_Shipping : 03 -- "ships" -> W_Package in G_Logistics
```

**Annotations (actor / group / sentence / work-object instance):**

```
domainstorytelling-beta

A_Customer : 01 -- "reports" -> W_Ticket id S_Report
A_Agent    : 02 -- "closes"  -> W_Ticket id S_Close

annotate actor A_Agent "Handles escalations."
annotate sentence 01 "Triggered by a customer request."
annotate workobject W_Ticket@S_Close "Closed ticket instance."
```

Activities that happen at the same time carry the same sequence number in this notation, so
numbers may repeat. The optional `id S_…` gives such sentences a stable annotation target.

**Full example — "Going to the movies" (Metropolis, coarse-grained, grouped by
subdomain — a sample from the Domain Storytelling literature):**

```
---
title: Metropolis - 1a Going to the movies—coarse-grained—grouped by subdomain
config:
  domainstorytelling:
    rankdir: TB
    nodeSpacing: 50
    rankSpacing: 100
---
domainstorytelling-beta

A_Cashier "Cashier"
A_Concession "Concession"
A_Moviegoer "Moviegoer"
A_Projectionist "Projectionist"
A_Usher "Usher"
W_Ticket "Ticket"
W_SnackAndDrinks "Snack & Drinks"
W_Movie "Movie"

group G_TicketSales "Ticket Sales" {
  A_Moviegoer : 01 -- "buys" -> W_Ticket -- "from" -> A_Cashier
}

group G_SnackSales "Snack Sales" {
  A_Moviegoer : 02 -- "buys" -> W_SnackAndDrinks -- "from" -> A_Concession
}

group G_Entrance "Entrance Control" {
  A_Moviegoer : 03 -- "shows" -> W_Ticket -- "to" -> A_Usher
  A_Usher : 04 -- "checks" -> W_Ticket -- "and grant entrance to" -> A_Moviegoer
}

group G_MovieProjection "Movie Projection" {
  A_Projectionist : 05 -- "starts" -> W_Movie -- "for" -> A_Moviegoer
  A_Moviegoer : 06 -- "watches" -> W_Movie
}
```

Common Mermaid features are supported: frontmatter `title`, `accTitle`/`accDescr`
accessibility, and per-diagram `config.domainstorytelling` layout options (`rankdir`,
`nodeSpacing`, `rankSpacing`, `diagramPadding`).

## Implementation

I will try and implement it myself.

### PR DESCRIPTION
**Pull request by bsydow**
## :bookmark_tabs: Summary

Adds **Domain Storytelling** as a new diagram type, available as `domainstorytelling-beta`.

Domain Storytelling is a collaborative modeling technique (Stefan Hofer & Henning Schwentner,
[domainstorytelling.org](https://domainstorytelling.org)) in which domain experts narrate
business processes as numbered sentences of the form *an **actor** does something with a
**work object** and passes it to another **actor***. This brings that notation into
Mermaid's text-based DSL, so domain stories become diff-able, reviewable, and
version-controlled alongside the code they describe.

Resolves #7942

### What it renders

- **Actors** and **work objects** as labeled icon nodes, drawn from a built-in notation icon
  pack (`person`, `people`, `system`, `document`, `folder`, `call`, `email`, `conversation`,
  `info`) with per-declaration override
- **Numbered activity arrows** forming a readable sentence sequence, with sequence-number
  circles placed along the arrows
- **Chained continuations** — multiple work objects/actors and reverse actor links in one
  sentence
- **Groups** incl. nesting and group blocks
- **Annotations** for actors, groups, sentences, and work-object instances, with
  bracket-shaped callout pointers
- **Optional sentence ids** (`id S_…`) as stable annotation targets, for stories where a
  sequence number is used more than once
- Frontmatter `title`, `accTitle`/`accDescr`, and `config.domainstorytelling` layout options

### Example

```
domainstorytelling-beta

A_Customer : 01 -- "places" -> W_Order
A_Service  : 02 -- "processes" -> W_Order
```

## :straight_ruler: Design Decisions

- **`-beta` suffix.** Ships as `domainstorytelling-beta`, matching Mermaid's convention for
  new diagram types, so the syntax can stabilize on community feedback before the suffix is
  dropped.
- **Langium grammar.** The parser lives in `@mermaid-js/parser`
  (`domainstorytelling.langium`), following the current new-diagram guidance rather than
  Jison. Type-prefixed identifiers (`A_`, `W_`, `G_`, `S_`) keep the grammar unambiguous and
  make each token self-describing in the source.
- **rendering-util pipeline.** The renderer is built on the newer `rendering-util`
  render pipeline with a dagre layout, exposing `rankdir` (`LR`/`RL`/`TB`/`BT`),
  `nodeSpacing`, `rankSpacing`, and `diagramPadding` via a schema-driven config.
- **Theme integration.** Actor, work-object, sequence-circle, and group colors adapt per
  theme via dedicated `domainstorytellingActorColor` /
  `domainstorytellingWorkobjectColor` / `domainstorytellingSequenceColor` /
  `domainstorytellingGroupColor` variables across the base, default, dark, forest, and
  neutral themes.
- **Built-in icons, no setup.** A `mermaid-domainstorytelling` pack ships with the diagram
  and is referenced by bare name, so a minimal story renders without any
  `registerIconPacks` call — `person` and `document` are the defaults. Arbitrary Iconify
  icons remain available in `pack:icon` form for anything beyond the base notation.
- **Repeated sequence numbers are allowed.** Activities that happen at the same time carry
  the same number in this notation, so the parser accepts repeats and only rejects an
  *annotation* that targets an ambiguous number — pointing at sentence ids instead.

### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://mermaid.js.org/community/contributing.html)
- [x] :computer: have added necessary unit/e2e tests.
- [x] :notebook: have added documentation. Make sure [`MERMAID_RELEASE_VERSION`](https://mermaid.js.org/community/contributing.html#update-documentation) is used for all new features.
- [x] :butterfly: If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

Adds `domainstorytelling-beta` as a new beta Mermaid diagram type for Domain Storytelling.

Highlights:
- End-to-end implementation: Langium grammar + parser (`packages/parser/src/language/domainstorytelling/*` and parser tests), stateless diagram DB (`DomainStorytellingDb`), renderer/styles (`packages/mermaid/src/diagrams/domainstorytelling/*`), and diagram detector wiring in `diagram-orchestration`.
- Diagram capabilities: actor/work-object nodes with labeled/icon support, ordered numbered activity steps (including reverse links), chained continuations, nested groups plus `{ ... }` group blocks, and bracket-style annotations for actors, groups, sentence targets, and work-object instances (including `W_...@S_...` targeting).
- Stable targeting: optional sentence IDs to disambiguate repeated sequence numbers for annotations.
- Config, metadata, accessibility: frontmatter (`title`, `accTitle`, `accDescr`) and `config.domainstorytelling` layout controls (`rankdir`, node/rank spacing, `diagramPadding`, `useMaxWidth`), including schema/types and generated config-defaults support.
- Theme integration: new theme variables (`domainstorytellingActorColor`, `domainstorytellingWorkobjectColor`, `domainstorytellingSequenceColor`, `domainstorytellingGroupColor`) added across theme implementations.
- Notation icon pack: built-in Iconify pack (`mermaid-domainstorytelling`) with default “person/document”-style icons, plus support for per-declaration icon overrides.
- Quality & docs: comprehensive parser/renderer unit tests (including sanitization/XSS assertions and annotation routing/newline handling), Cypress visual snapshot tests (including ELK layout mode), multi-diagram SVG ID uniqueness coverage, new syntax/docs pages, updated examples/demo pages, cspell term updates, and a changeset for a minor release enabling the new diagram type.

### CODE REVIEW

File: cypress/integration/rendering/domainstorytelling/domainstorytelling.spec.ts [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    imgSnapshotTest(
+      `domainstorytelling-beta
+group G_Service "Service Desk"
+A_Agent "Support Agent" in G_Service
+A_Customer "Customer"
+
+A_Customer : 01 -- "reports" -> W_Ticket id S_ReportTicket
+A_Agent : 02 -- "triages" -> W_Backlog id S_Triage
+A_Agent : 03 -- "updates" -> W_Ticket id S_UpdateTicket
+A_Agent : 03 -- "closes" -> W_Ticket id S_CloseTicket
```
[the remark this answers is not in the document]
Comment (bsydow) [reply cypress/integration/rendering/domainstorytelling/domainstorytelling.spec.ts:103]: Intentional. In Domain Storytelling, activities that happen at the same time - or whose order doesn't matter - carry the same sequence number, so repeats are part of the notation rather than a slip.

**Comment by pbrolin47** [comment]
Hi @bsydow 
Thanks for the work you invested in this PR. 

1. I run the dev-server locally and checked the rendering and looked at the Argos-tests. The rendering looks fine, but I have minor suggestion. When looking at some of the examples in https://domainstorytelling.org/quick-start-guide I see that the Activities-arrow  pointing to an object is oriented so it points to the icon/image of the object. In this PR, the arrows is pointing between the icon/image and label. Not a big thing, but just a suggestion


image

image


After your fixes of the last review by @knsv-bot, I ran a review.

[sisyphos-bot]

 ---
  What's working well
   
  🎉 Plugin contract is complete and correct. diagram-orchestration.ts registers the detector, domainstorytellingDiagram.ts uses get db() { return new DomainStorytellingDb(); } for fresh-per-render
  isolation, clear() resets all arrays and maps including pendingGroupMemberships. Exactly right.

  🎉 Continuation model is elegant. The three continuation types — AdditionalWorkObject, AdditionalActor, ReverseActor — cleanly model the directional flow of domain stories. The latestWorkobjectId
  pointer in processSentence follows a well-documented invariant (see the comment on ReverseActor anchoring). The 30+ parser spec cases validate the edge cases thoroughly.

  🎉 XSS coverage is proactive. The renderer spec (domainstorytellingRenderer.spec.ts) includes five dedicated XSS tests covering actor/workobject labels, annotation bodies, edge labels, and sentence
  annotations — all using real attack payloads. sanitizeText() is called before every HTML label build in addVertices and addEdges. This is the right defense-in-depth pattern.

  🎉 Sequence circle collision detection. The three-phase slide algorithm (sample → slide until clear → draw) keeps badges on their own connectors rather than stacking. The containerNode-scoped
  tracking handles nested cluster subgraphs correctly.

  🎉 Theme integration. All five themes define all four domainstorytelling* variables. theme-base.js derives defaults from primaryBorderColor/tertiaryBorderColor/secondaryBorderColor, so new theme
  variants get sensible colors for free.

  ---
  Things to fix
   
  🟡 [important] arrowheadStyle: 'fill: #333' hardcodes a dark color (domainstorytellingRenderer.ts, addEdges)

  arrowheadStyle: 'fill: #333',
   
  This inline style is applied directly to SVG arrowhead markers, bypassing the CSS theme system. In dark/forest/neutral themes the arrowhead fill stays #333 (near-black), which will be invisible or
  very low-contrast against dark backgrounds. The edge stroke is correctly theme-aware via CSS (.domainstorytelling-link { stroke: ${options.lineColor ?? '#333'} }), but arrowheads aren't styled by
  that class.
 
  Fix: remove the hardcoded value and let the marker inherit from the edge stroke color, or pass a theme-derived value. Other Mermaid renderers that use the render() pipeline typically omit
  arrowheadStyle or derive it from getConfig(). Check the flowchart renderer for the established pattern.

  ---
  🟡 [important] Stale copy-paste comment in domainstorytellingTypes.ts
   
  export interface DomainstorytellingDB extends DiagramDB {
    // config
    //getConfig: () => Required<PieDiagramConfig>;   ← references the pie diagram

  This was left from whichever diagram was used as a template. Remove it.

  ---
  Nits (non-blocking)
   
  🟢 SEQNO grammar only accepts two-digit+ sequences. /0[1-9]|[1-9][0-9]+/ matches 01-09 and 10+, but not single-digit 1-9. A user writing : 5 -- "verb" -> W_W will get a parse error. This looks
  intentional (all examples and tests consistently use 01, 02 style), but it's worth a sentence in the docs or an error message that suggests the correct format.

  🟢 Theme color consistency. theme-base.js derives colors from existing theme primitives (primaryBorderColor, etc.), which is ideal. The four concrete theme files hardcode RGB hex values. This
  follows the existing cynefin/radar pattern, so it's not wrong — but for future-proofing consider using khroma-adjusted derivations the way theme-base.js does.

  ---

**Comment by bsydow** [comment]
Hi @pbrolin47,

Thanks for the review! I fixed the arrow orientation. They now point at the icon instead of between the icon and label, like the [quick-start guide](https://domainstorytelling.org/quick-start-guide) examples.

**Comment by bsydow** [comment]
Hi @pbrolin47 

All four points have been addressed:

**Hardcoded arrowhead colour** — removed. The arrowhead now inherits the theme's line colour through the shared marker styling, so it matches the edge stroke on every theme.

**Stale copy-paste comment** — removed, along with a check that no other template leftovers remain in the diagram.

**Single-digit sequence numbers** — fixed rather than documented. The restriction turned out to be unnecessary, so the terminal now accepts 1–9 alongside the zero-padded form; both parse to the same number and render identically. 0 and 00 remain rejected, since a falsy sequence number would silently lose its label at render time. Docs and tests updated.

**Theme colour derivation** — not adopted, but the point was worth raising: chasing it surfaced a real bug. Measured against the theme's border variables, deriving regresses contrast on every theme and collapses actor and work object to the same colour on one of them, because those variables are desaturated outlines rather than category colours. What the investigation did turn up is that only 5 of the 11 registered themes defined these colours at all; the other 6 emitted color: undefined into the stylesheet, an invalid declaration browsers drop silently, so icons lost their colour with nothing in the console. Unit tests only ever rendered the default theme and the e2e spec pins none, so it went unseen. All 11 themes now carry values, the stylesheet has a fallback for any theme added later, and a test covers both across every registered theme. The base theme, which did derive, gained a large contrast improvement as a side effect. Separately, the per-diagram colour overrides were already supported but undocumented — there is now a Styling section covering them.
